### PR TITLE
common : refactor arg parser

### PR DESCRIPTION
**Pull request by ngxson**
## TL;DR for breaking change

This PR has only some small breaking changes to the **environment variable** system introduced in #9105 : 
- To disable continuous batching, add `LLAMA_ARG_NO_CONT_BATCHING=1` (instead of `LLAMA_ARG_CONT_BATCHING=0`)
- To disable slots endpoint, add `LLAMA_ARG_NO_ENDPOINT_SLOTS=1` (instead of `LLAMA_ARG_ENDPOINT_SLOTS=0`)
- If both command line argument and environment variable are both set for the same param, the **argument** will take precedence over env var.  
  In such case, you will also see a warning, for example: `warn: LLAMA_ARG_CTX_SIZE environment variable is set, but will be overwritten by command line argument -c`

## In this PR

The goals of this PR are:
- Refactor & make the arg parser code more intuitive, tailored to llama.cpp's usage
- Ability to auto-generate documentation (markdown content) from code
- Better support for multiple examples (plus, one arg used by multiple examples or different purposes)
- Unifying env variable & arguments logic into one place

To generate markdown, run: `make llama-export-docs`  
Output file will be in the format of `autogen-{EXAMPLE_NAME}.md`

TODO:
- [x] migrate all args to this new format
- [x] migrate env variables
- [x] add some ctests
- [x] add binary target to export markdown
- [x] update exiting markdown docs

---

List of removed args (this is **not** breaking change, since no where in the code base handles these args):

```
--priority
--priority-batch
--priority-draft
-Cbd
--priority-batch-draft
```

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Medium

### CODE REVIEW

**Comment by ngxson** [comment]
Before proceeding further, I would like to ask for you opinion about this subject. Do you think this is a good way to have code-as-documentations? And if so, do you have any idea to add to the implementation? Thank you.

**Comment by ngxson** [comment]
@ggerganov  Before proceeding further, I would like to ask for you opinion about this subject. Do you think this is a good way to have code-as-documentations? And if so, do you have any idea to add to this? Thank you.

File: common/common.h [hunk 1]
```diff
@@ -277,6 +278,66 @@ struct gpt_params {
...
+    llama_arg(std::vector<std::string> args, std::string help, std::function<bool(std::string)> handler) : args(args), help(help), handler_string(handler) {}
+
+    llama_arg(std::vector<std::string> args, std::string help, std::function<bool(bool)> handler) : args(args), help(help), handler_bool(handler) {}
+
+    llama_arg(std::vector<std::string> args, std::string help, std::function<bool(void)> handler) : args(args), help(help), handler_void(handler) {}
+
+    llama_arg & set_examples(std::set<enum llama_example> _examples) {
+        examples = std::move(_examples);
+        return *this;
+    }
```
Comment (ggerganov) [inline common/common.h:308]: Avoid identifiers with leading underscore:

```suggestion
    llama_arg & set_examples(std::set<enum llama_example> examples) {
        this->examples = std::move(examples);
        return *this;
    }
```

**Review by ggerganov — COMMENTED** [review]
Seems ok to me 👍

**Comment by ngxson** [comment]
@ggerganov Thank you for the initial review. This PR is now ready.

Here is a quick recap on what I've done:
- All CLI args & env var are migrated to the new format
- `tests/test-arg-parser` is added to test this new system
- `llama-export-docs` target is added, which can export list of arguments to markdown table (instead of markdown list in my initial demo). [Here is an example for server docs](https://github.com/ngxson/llama.cpp/blob/xsn/argparser_v3/examples/server/README.md#usage)

File: tests/test-arg-parser.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+#undef NDEBUG
+#include <cassert>
+
+#include "common.h"
+
+int main(void) {
+#ifdef _WIN32
+    printf("test-arg-parser: skip on windows build\n");
+#else
```
Comment (slaren) [inline tests/test-arg-parser.cpp:13]: What's the reason for skipping this on windows? A brief comment would be useful.

File: tests/test-arg-parser.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    setenv("LLAMA_ARG_MODEL", "blah.gguf", true);
+    setenv("LLAMA_ARG_THREADS", "1010", true);
+    argv = {"binary_name", "-m", "overwritten.gguf"};
+    assert(true == gpt_params_parse(argv.size(), list_str_to_char(argv).data(), params, options));
+    assert(params.model == "overwritten.gguf");
+    assert(params.cpuparams.n_threads == 1010);
+
+
+    printf("test-arg-parser: all tests OK\n\n");
+#endif // __MINGW32__
```
Comment (slaren) [inline tests/test-arg-parser.cpp:94]: The command is wrong (should be `_WIN32`).

File: tests/test-arg-parser.cpp (hunk 2 above)
Comment (ngxson) [reply tests/test-arg-parser.cpp]: Yeah I forgot to add a comment. Indeed, the test failed to compile because `setenv` is not supported on windows. I also exclude only the `setenv` part of the test: https://github.com/ggerganov/llama.cpp/pull/9308/commits/d545ffcb6deac469a0b4af0be69bf9b47997181e

**Comment by ggerganov** [comment]
Functionality-wise, this is great. However the build time of `libcommon` increases on my machine from ~3s to ~12s:

```bash
ccache -C && touch ../common/common.cpp && time make -j common
```

Should we try to reduce it in some way? I suppose the culprit is in the lambda handlers.

**Comment by slaren** [comment]
Yeah this is pretty bad. 18 seconds to compile on 13900k.

**Comment by ngxson** [comment]
Thanks for testing that. Yes I can confirm that the build time is now ~9.6s compared to ~5.8s on master (using macbook M3 Max).

Apply the patch by @slaren bring it down to 7.8s, which is exactly 2s slower than master.

Testing a bit further, I changed all the `std::vector<std::string> & args` to `std::string & args` and it brought the build time back to 5.2s

So at this point I'm a bit doubt if I can somehow take advantage of this to reduce the build time, without compromising the runtime performance. In worst case, what is an acceptable increase in build time?

**Comment by slaren** [comment]
> Testing a bit further, I changed all the `std::vector<std::string> & args` to `std::string & args` and it brought the build time back to 5.2s

I am a bit confused by this, do you mean the `args` member in `llama_arg`? Are there other vectors like this?

**Comment by ngxson** [comment]
> I am a bit confused by this, do you mean the `args` member in `llama_arg`? Are there other vectors like this?

Yes it is.

The way I tested was:
1. Remove all code and leaving only one of the arg
2. Repeat the arg 190 times
3. Try compiling. Even though it's one arg repeated 190 times, it will still take the same time as having 190 different args (just to prove that there is nothing being cached)
4. Now change the first arg from std::vector to a simple std::string
5. Re-compile, now it should take much less time

**Comment by slaren** [comment]
Ok, I see. Replacing the vectors in the `llama_arg` constructors with `std::initializer_list<const char*>` should improve the build time a bit without too many changes.

**Comment by ngxson** [comment]
> I am a bit confused by this, do you mean the `args` member in `llama_arg`? Are there other vectors like this?

I mean the `args` in constructor of `llama_arg`

The way I tested was:
1. Remove all code and leaving only one of the arg (I took "--verbose" in my case)
2. Repeat the arg 190 times
3. Try compiling. Even though it's one arg repeated 190 times, it will still take the same time as having 190 different args (just to prove that there is nothing being cached)
4. Now change the first arg from std::vector to a simple std::string
5. Re-compile, now it should take much less time

In the end, the contructor becomes (for testing purpose, I save the string to `env` - the code won't work, but just to test the built time):

```
llama_arg(const std::string args, const std::string & value_hint, const std::string & help, void (*handler)(gpt_params & params, llama_sampling_params & sparams, int)) : env(args), value_hint(value_hint), help(help), handler_int(handler) {}
```

**Comment by ngxson** [comment]
> Ok, I see. Replacing the vectors in the `llama_arg` constructors with `std::initializer_list<const char*>` should improve the build time a bit without too many changes.

Wow not just a bit, it's now back to 5.2s. Thanks for the hint about `std::initializer_list`, I didn't know about that 😄

**Comment by ngxson** [comment]
Quick question: because `sparams` is already included inside `params`, should we get rid of passing `sparams` as argument of handler function? (and replace it with `params.sparams`

**Comment by ngxson** [comment]
Alright, I change some `std::string` to `const char *` and it brought down the build time a little bit. Keep in mind that the 5.2s result above was with only one arg repeated 190 times.

The build time of the latest commit:

```
$ make clean && time make -j common/common.o
make -j common/common.o  5.70s user 0.16s system 99% cpu 5.889 total

# versus master
# make -j common/common.o  5.48s user 0.16s system 99% cpu 5.674 total
```

So it's +0.2s compared to master. It can be further brought down if `std::string help` can be changed to `const char *`, but currently it's a quite tricky to have `format(...)` function to return `const char *`, so I think we can consider it later.

File: examples/export-docs/export-docs.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#include "common.h"
+#include "llama.h"
+
+#include <cmath>
+#include <cstdio>
+#include <fstream>
+#include <string>
+#include <vector>
+
```
Comment (ggerganov) [inline examples/export-docs/export-docs.cpp:9]: Remove the unused headers

File: examples/export-docs/export-docs.cpp
Comment (ggerganov) [inline examples/export-docs/export-docs.cpp:1]: I guess `gen-docs` would be a better name, but `exposr-docs` also works

File: common/common.h [hunk 5]
```diff
@@ -280,13 +303,93 @@ struct gpt_params {
...
+
+// parse input arguments from CLI
+// if one argument has invalid value, it will automatically display usage of the specific argument (and not the full usage message)
+bool gpt_params_parse   (int argc, char ** argv, gpt_params & params, std::vector<llama_arg> & options);
+bool gpt_params_parse_ex(int argc, char ** argv, gpt_params & params, std::vector<llama_arg> & options);
+
+// print full usage message; it will be called internally by gpt_params_parse() if "-h" is set
+void gpt_params_print_usage(gpt_params & params, std::vector<llama_arg> & options);
+
+void gpt_params_handle_model_default(gpt_params & params);
```
Comment (ggerganov) [inline common/common.h:389]: No need to be public

**Review by ggerganov — APPROVED** [review]

File: common/common.h [hunk 6]
```diff
@@ -276,13 +300,93 @@ struct gpt_params {
     bool batched_bench_output_jsonl = false;
 };
 
-void gpt_params_parse_from_env(gpt_params & params);
-void gpt_params_handle_model_default(gpt_params & params);
+struct llama_arg {
```
Comment (ggerganov) [inline common/common.h:303]: Should move the method implementations in the .cpp, to avoid building the same code in all examples.

Can also move all `llama_arg` related stuff in `common/arg.h,.cpp`. Can be in a follow-up PR

Comment (ngxson) [reply common/common.h:303]: Currently doing this will break `gen-docs`, since it reads the data directly from class member.

But yes I will do a follow up to split it into `common/arg.h,.cpp`

File: examples/export-docs/export-docs.cpp (hunk 4 above)
Comment (ngxson) [reply examples/export-docs/export-docs.cpp]: Done in https://github.com/ggerganov/llama.cpp/pull/9308/commits/4b96c69a08a0e0e6f09883f7d8bad1375fbcaf86

File: examples/export-docs/export-docs.cpp
Comment (ngxson) [reply examples/export-docs/export-docs.cpp:1]: Yup that make sense. I changed it to `gen-docs`: https://github.com/ggerganov/llama.cpp/pull/9308/commits/4b96c69a08a0e0e6f09883f7d8bad1375fbcaf86

File: common/common.h (hunk 5 above)
Comment (ngxson) [reply common/common.h]: Done in https://github.com/ggerganov/llama.cpp/pull/9308/commits/4b96c69a08a0e0e6f09883f7d8bad1375fbcaf86
